hooks/hook.go

```diff
@@ -0,0 +1,40 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// The hooks package contains all hook runners supported by snapd.
```

Comment:
If we hold them all here this will end up very cross-concern. My initial idea was to try to put the common pieces as part of the public interface of the hookstate package, and then have the actual implementation of the hooks inside the managers responsible for their respective logic (interface hooks in ifacestate, config hooks in configstate, etc).

For that, we'll need to allow these managers to register the fact they may be called. First cut might look along those lines, perhaps:
```
hookstate.Register("^some-hook-name$", hookHandler)

func hookHandler(*hookstate.Context) hookstate.Handler { ... }

type Context struct {
        HookState
        PID int
}

func (c *Context) Kill() error

type Handler interface {
        Before() error
        Done() error
        Error(err error) error
}
```